docs/animation/asm.md

```diff
@@ -0,0 +1,312 @@
+    }
+    else {
+        asm = null;
+    }
+}
+```
+
+Here, an asm is loaded with one extra parameter, named `cycle_length`. This parameter is of the type `VariableValue`, so we can
+set it from within our code.
+
+Using an ASM instance, you can get the current state with `.currentState()` and transition to another state with `.transition(nextState)`
+ 
+`VariableValue` parameters can have their value set by calling `.setValue`, but oddly you can not read this value back. There is no need
+to inform the ASM of this change, it happens automatically. All of the other clips can be used from code too if you so desire, but I will not
+go into their API here as they are mainly designed to be used in the JSON files.
+
+File Format
+-------------
+
+The ASMs have a slightly strange and confusing file format, but here's how it works.
+
+First, a simple example:
+```json
+{
+  "parameters": {
+    "anim_cycle": ["/", "#cycle_length"]
+  },
+  "clips": {
+    "default": ["apply", "forgedebugmodelanimation:block/rotatest@default", "#anim_cycle" ]
+  },
+  "states": [
+    "default"
+  ],
+  "transitions": {},
+  "start_state": "default"
+}
+```
+
+As stated above, the files have parameters, clips, states and transitions, as well as the starting state.
+All of these tags are required, even if they are empty.
+
+### Parameters
+
+```javascript
+{
+    "name": <parameter_definition>
+}
+```
+
+Different types of parameters have different formats for `<parameter_definition>`, and the simple ones are:
+
+- `IdentityValue`: the string `#identity`,
+- `ParameterValue`: the parameter to reference, prefixed with `#`, e.g. `#my_awesome_parameter`
+- `ConstValue`: a number to use as the constant to return
+
+Other kinds are a bit more complicated
+
+#### `SimpleExprValue`
+
+Format: `[ regex("[+\\-*/mMrRfF]+"), <parameter_definition>, ... ]`
+
+##### Examples:
+```json
+[ "+", 4 ]
+[ "/+", 5, 1]
+[ "++", 2, "#other" ]
+[ "++", "#other", [ "compose", "#cycle", 3] ]
+```
+
+##### Explanation
+
+`SimpleExprValue` takes its input and applies operations to it.
+The first value in the list are the operations to apply. Each operation is applied in order to the input.
+Each following value is another parameter definition which is evaluated with the unmodified input value.
+
+##### Operations (case-sensitive):
+
+- `+`: output = input + arg
+- `-`: output = input - arg
+- `*`: output = input * arg;
+- `/`: output = input / arg;
+- `m`: output = min(input, arg);
+- `M`: output = max(input, arg);
+- `r`: output = floor(input / arg) * arg;
+- `R`: output = ceil(input / arg) * arg;
+- `f`: output = input - floor(input / arg) * arg;
+- `F`: output = ceil(input / arg) * arg - input;
+
+##### Example explanations:
+- input + 4
+- (input / 5) + 1
+- input + 2 + value of parameter other
+- input + value of parameter other + value of parameter cycle given input 3
```

Comment:
I'm reading this md file unrendered but maybe move these to be right under their respective examples?

Reply:
i think it might be better this was, so you can see some examples, get the explanation of the entire parameter type, and then you can check to see the explanations of the examples

Reply:
The idea I had with the examples was to first show examples, then explain the parameter, and then show the examples again with their meaning now that the reader has read the explanation.